src/containerapp/azext_containerapp/containerapp_decorator.py

```diff
@@ -721,3 +725,92 @@ def set_up_scale_rule(self):
             scale_def["rules"] = [scale_rule_def]
 
         return scale_def
+
+
+class ContainerAppPreviewCreateDecorator(ContainerAppCreateDecorator):
+    def __init__(
+        self, cmd: AzCliCommand, client: Any, raw_parameters: Dict, models: str
+    ):
+        super().__init__(cmd, client, raw_parameters, models)
+
+    def construct_containerapp(self):
+        super().construct_containerapp()
+        self.set_up_extended_location()
+
+    def set_up_extended_location(self):
+        if self.get_argument_environment_type() == CONNECTED_ENVIRONMENT_TYPE:
+            if not self.containerapp_def.get('extendedLocation'):
+                parsed_env = parse_resource_id(self.get_argument_managed_env())  # custom_location check here perhaps
+                env_name = parsed_env['name']
+                env_rg = parsed_env['resource_group']
+                env_info = self.get_environment_client().show(cmd=self.cmd, resource_group_name=env_rg, name=env_name)
+                self.containerapp_def["extendedLocation"] = env_info["extendedLocation"]
+
+    def get_environment_client(self):
+        if self.get_argument_yaml():
+            env = safe_get(self.containerapp_def, "properties", "environmentId")
+        else:
+            env = self.get_argument_managed_env()
+
+        environment_type = self.get_argument_environment_type()
+        if not env and not environment_type:
+            return ManagedEnvironmentClient
+
+        parsed_env = parse_resource_id(env)
+
+        # Validate environment type
+        if parsed_env.get('resource_type').lower() == CONNECTED_ENVIRONMENT_RESOURCE_TYPE.lower():
+            if environment_type == MANAGED_ENVIRONMENT_TYPE:
+                logger.warning("User passed a connectedEnvironment resource id but did not specify --environment-type connected. Using environment type connected.")
+            environment_type = CONNECTED_ENVIRONMENT_TYPE
+        else:
+            if environment_type == CONNECTED_ENVIRONMENT_TYPE:
+                logger.warning("User passed a managedEnvironment resource id but specified --environment-type connected. Using environment type managed.")
+            environment_type = MANAGED_ENVIRONMENT_TYPE
+
+        self.set_argument_environment_type(environment_type)
+        self.set_argument_managed_env(env)
+
+        if environment_type == CONNECTED_ENVIRONMENT_TYPE:
+            return ConnectedEnvironmentClient
+        else:
+            return ManagedEnvironmentPreviewClient
+
+    def get_argument_environment_type(self):
+        return self.get_param("environment_type")
+
+    def set_argument_environment_type(self, environment_type):
+        self.set_param("environment_type", environment_type)
+
+
+class ContainerAppPreviewListDecorator(BaseContainerAppDecorator):
```

Comment:
why do we need listDecorator? can these operations merged to one decorator?

Reply:
`list` command need to support preview argument `environment-type`.